### PR TITLE
Unit test for PsFacebook

### PR DESCRIPTION
**Pull request by d-molotchko**
Fixed issue #41 
1. Added unit test for PsFacebook
2. Refactor PsFacebook class 
 * Added additional ctor with Request and Response for avoiding facebook collaboration
 * Extracted 'facebookClient' and 'token' functions for request\response processing
 * Added @checkstyle suppression for Class Data Abstraction Coupling
 * Added puzzle that will try to decouple this class.

### CODE REVIEW

**Comment by davvd** [comment]
Thanks, I will find someone to review this PR soon

**Comment by davvd** [comment]
@ggajos please review, thanks

**Comment by ggajos** [comment]
@d-molotchko please add simple description to pull request, what do you actually implementing here and why you do this that way.

**Comment by ggajos** [comment]
@d-molotchko I would suggest to skip this warning for now using @checkstyle suppression and add puzzle that will try to decouple this class.

File: src/test/java/org/takes/facets/auth/social/PsFacebookTest.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * @since 0.10.2
+ */
+public final class PsFacebookTest {
+
+    /**
+     * PsFacebook can return identity iterator.
+     * @throws IOException If some problem inside
+     */
+    @Test
+    public void returnIdentityIterator() throws IOException {
```
Comment (ggajos) [inline src/test/java/org/takes/facets/auth/social/PsFacebookTest.java]: let's rename to `returnsIdentityIterator`
http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names

**Comment by ggajos** [comment]
@d-molotchko few comments

**Comment by d-molotchko** [comment]
@ggajos done. I'll create puzzle after PR merging

**Comment by ggajos** [comment]
@d-molotchko the idea of puzzles is to put them into the code with PR, why you would like to do it separately?

**Comment by d-molotchko** [comment]
@ggajos I added puzzle for class decoupling

**Comment by ggajos** [comment]
@d-molotchko build is failing due to:
```
[WARNING] Used undeclared dependencies found:
	com.jcabi:jcabi-immutable:jar:1.3:compile
	javax.validation:validation-api:jar:1.1.0.Final:compile
```
I think you should add those dependencies to pom.xml if you explicitly using those classes

**Comment by d-molotchko** [comment]
@ggajos thanks, it was my mistake. I remove those dependencies

**Comment by ggajos** [comment]
@rultor merge please

File: src/main/java/org/takes/facets/auth/social/PsFacebook.java [hunk 2]
```diff
@@ -66,13 +78,51 @@
...
+    private final transient Map.Entry<com.jcabi.http.Request,
+        WebRequestor.Response> testing;
+
+    /**
      * Ctor.
      * @param fapp Facebook app
      * @param fkey Facebook key
      */
     public PsFacebook(final String fapp, final String fkey) {
+        this(fapp, fkey, null);
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/social/PsFacebook.java]: see http://www.yegor256.com/2014/05/13/why-null-is-bad.html we don't use NULL, ever

File: src/main/java/org/takes/facets/auth/social/PsFacebook.java [hunk 3]
```diff
@@ -126,6 +176,42 @@ private static User fetch(final String token) {
     }
 
     /**
+     * Get Facebook client.
+     * @param token Facebook access token
+     * @return The Facebook client implementation
+     */
+    private DefaultFacebookClient facebookClient(final String token) {
```
Comment (yegor256) [inline src/main/java/org/takes/facets/auth/social/PsFacebook.java]: see http://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html

**Comment by yegor256** [comment]
@d-molotchko you're on the right track, but there are a few design issues. don't use NULL, ever. it's very important.

**Comment by ggajos** [comment]
@d-molotchko any update here?

**Comment by d-molotchko** [comment]
@ggajos no, sorry

**Comment by ggajos** [comment]
@d-molotchko I see that task has been reassigned, we can close this PR?

**Comment by ggajos** [comment]
@d-molotchko think you should close this PR

**Comment by d-molotchko** [comment]
@ggajos ok

**Comment by davvd** [comment]
@elenavolokhova please, review this ticket for compliance with our [QA rules](http://www.teamed.io/qa.html)

**Comment by elenavolokhova** [comment]
@ggajos 
>davvd commented 27 days ago
Thanks, I will find someone to review this PR soon

Not all comments are addressed here.
Could you please confirm that you'll pay attention to this in future?

**Comment by ggajos** [comment]
@davvd please address you comments directly

**Comment by ggajos** [comment]
@elenavolokhova yes, thanks, I overlooked this

**Comment by elenavolokhova** [comment]
@ggajos Thank you!

**Comment by elenavolokhova** [comment]
@davvd Quality is acceptable.

**Comment by davvd** [comment]
> @davvd Quality is acceptable.

@elenavolokhova thanks for the QA review, we'll work better next time

**Comment by davvd** [comment]
@ggajos 10 mins was added to the account of @elenavolokhova (for QA review), in transaction 56371594... thanks, I just added **26 mins** to your account, payment `AP-7U787004C1185373A`, 504 hours and 36 mins spent... extra minutes for review comments (c=11)... +26 added to your rating, at the moment it is: [+3300](http://www.netbout.com/b/35094?open=rating)

**Pull request closed without merging**